eclipse-zenoh/zenoh
# Mutual authentication

# Features

* Upgraded the async-rustls library, now using  rustls 0.20
* Adapted the zenoh-link-tls package to work with this new version of async-rustls
* Added two way authentication if configured that way.

# How to test the 2 way authentication
Let's create 2 certificates, one for the "server" and one for the "client". For this, it is explained in the zenoh documentation regarding TLS (https://zenoh.io/docs/manual/tls/) how to create certificates using MiniCA.

You can store the certificates as follows, under a certificates folder on your zenoh folder. It should look like this:

```
certificates
├── client
│   ├── localhost
│   │   ├── cert.pem
│   │   └── key.pem
│   ├── minica-key.pem
│   └── minica.pem
└── server
    ├── localhost
    │   ├── cert.pem
    │   └── key.pem
    ├── minica-key.pem
    └── minica.pem
```
From your zenoh folder, after having build this branch,

1- run the router with: `RUST_LOG=debug cargo run --bin=zenohd -- -c configs/router.json5` 

2- run a publisher with: `RUST_LOG=debug ./target/debug/examples/z_pub -c configs/client.json5`

3- run a subscriber with: `RUST_LOG=debug ./target/debug/examples/z_sub -c configs/client.json5`

where `configs/client.json5` is
```
{
  /// The node's mode (router, peer or client)
  mode: "client",
  connect: {
    endpoints: [ "tls/localhost:7447" ]
  },
  transport: {
    link: {
      tls: {
        root_ca_certificate: "./certificates/client/minica.pem",
        client_auth: true,
        client_private_key: "./certificates/server/localhost/key.pem",
        client_certificate: "./certificates/server/localhost/cert.pem",
      },
    },
  },
}
```

and `configs/router.json5` is:

```
{
  /// The node's mode (router, peer or client)
  mode: "router",
  listen: {
    endpoints: [ "tls/localhost:7447" ]
  },
  transport: {
    link: {
      tls: {
        client_auth: true,
        root_ca_certificate: "./certificates/server/minica.pem",
        server_private_key: "./certificates/client/localhost/key.pem",
        server_certificate: "./certificates/client/localhost/cert.pem",
      },
    },
  },
}
```

(see the configs are inside a folder `configs` under your `/zenoh` folder).
